drivers/hud_spm_report/app/models/hud_spm_report/fy2023/spm_enrollment.rb

```diff
@@ -0,0 +1,280 @@
+      end
+    end
+
+    private_class_method def self.total_income(income_benefit)
+      income_benefit&.hud_total_monthly_income&.clamp(0..) || 0
+    end
+
+    private_class_method def self.earned_income(income_benefit)
+      (income_benefit&.earned_amount || 0).clamp(0..)
+    end
+
+    private_class_method def self.non_employment_income(income_benefit)
+      (total_income(income_benefit) - earned_income(income_benefit)).clamp(0..)
+    end
+
+    private_class_method def self.current_income_benefits(enrollment, end_date)
+      # Exit assessment for leavers, or most recent annual update within report range for stayers
+      if enrollment.exit.present? && enrollment.exit.exit_date <= end_date
+        enrollment.income_benefits_at_exit
+      else
+        enrollment.
+          income_benefits_annual_update.
+          where(information_date: ..end_date).
+          where(annual_update_window_sql(enrollment)).
+          order(information_date: :desc).
+          first
+      end
+    end
+
+    private_class_method def self.previous_income_benefits(enrollment, annual_date, end_date)
+      return enrollment.income_benefits_at_entry if enrollment.exit.present? && enrollment.exit.exit_date <= end_date
+
+      # Most recent annual update on or before the renewal date, or the entry assessment
+      enrollment.
+        income_benefits_annual_update.
+        where(information_date: ...annual_date).
+        where(annual_update_window_sql(enrollment)).
+        order(information_date: :desc).
+        first || enrollment.income_benefits_at_entry # Default to entry assessment if less than 2 years
+    end
+
+    private_class_method def self.annual_update_window_sql(enrollment)
+      # 30 days of anniversary of entry date
+      report_date = ib_t[:information_date].to_sql
+      entry_date = enrollment.entry_date.to_s(:db)
+      interval = '30 days'
+      <<~SQL
+        (EXTRACT(MONTH FROM #{report_date}), EXTRACT(DAY FROM #{report_date})) IN (
+          SELECT EXTRACT(MONTH FROM gs), EXTRACT(DAY FROM gs)
+          FROM generate_series(
+              '#{entry_date}'::date - INTERVAL '#{interval}',
+              '#{entry_date}'::date + INTERVAL '#{interval}',
+              '1 day'
+          ) AS gs
+        )
+      SQL
+    end
+
+    def self.enrollment_own_move_in_date(enrollment)
+      return nil unless enrollment.move_in_date
+
+      enrollment.move_in_date >= enrollment.entry_date ? enrollment.move_in_date : nil
+    end
+
+    private_class_method def self.household(enrollments)
+      result = {}
+      scope = enrollments.heads_of_households
+      scope.find_in_batches do |batch|
+        batch.each do |enrollment|
+          result[enrollment.household_id] = HomelessnessInfo.new(
+            start_of_homelessness: enrollment.date_to_street_essh,
+            entry_date: enrollment.entry_date,
+            move_in_date: enrollment_own_move_in_date(enrollment),
+          )
+        end
+      end
```

Comment:
Pretty sure this is going to enforce 1 HoH per household (which is fine) but it's going to pick the HoH randomly based on the DB return order.  We may want to note that.  Or maybe we want to order it more explicitly.  Since we're looking for move-in dates, I'd be really tempted to 
```
.order(:household_id, :move_in_date)
```
As long as we can force nulls last.  That way if one HoH has the move-in date we'd be using it.